### DISCUSSION TITLE
FQDN references can't resolve for MQTT broker host

### DISCUSSION THREAD

**Discussion by dachshund-digital** [arendst/Tasmota#12111] [Q&A] [answered]
Tell me why FQDN references to not resolve for MQTT broker host specification? If I have a DNS server. If I have proper FQDN resolution for host names, Tasmota should resolve said names, period. That is how everything works with DNS. But what Tasmota does is some thing other than the standard default methodology? I don't use or want to use SD-DNS, I don't use or want to use mDNS, I have real DNS servers, that are entire level standard. Why, I happen to be a 30 IT retired network engineer, so creating my own DNS infrastructure was painless. I just don't see the logic of not following standards.

**Comment by s-hadinger** [comment, marked as answer]
Tasmota uses dns in a standard way. mdns is disabled by default. What is your problem?

**Reply by sfromis** [reply]
Users opting for a public MQTT broker host could not work if the FQDN was not resolved.

**Comment by dachshund-digital** [comment]
Then what I am seeing makes no sense.   I have a master/slave DNS server setup, the slave is caching only.  DNS in all respects appears to be working correctly.  The MQTT broker is defined in DNS forward and reverse zones, as pi4modelb1.dd.org (local domain private domain).  A CNAME record as broker.dd.org points back to pi4modelb1.dd.org.  If I enter either the actual name or the CNAME name, the tasmota (9.4) based ESP8266 device fails to connect to the broker.  If I enter the IP address the connection is successful.  Testing the DNS name and CNAME records, everything appears in order.  using dig or nslookup, all works as expected, it is only the tasmota based ESP8266 device that fails to connect to the MQTT broker my name.  Watching the COM port output of tasmota console, confirms the failure.  For example   

`
03:12:34.046 MQT: Attempting connection...
03:12:34.282 MQT: Connect failed to pi4modelb1.dd.org:1883, rc -2. Retry in 60 sec
`

Whereas using the IP address works...

`03:13:38.092 MQT: MqttHost 192.168.1.102, MqttPort 1883, MqttClient tasmota_840D8E8FF512, MqttUser [redacted], Topic tasmota_840D8E8FF512, FullTopic %prefix%/%topic%/
`

So I say again, if what you say is true... which I do believe you, this makes no sense at yet.

**Reply by s-hadinger** [reply]
I can't test right now but this looks strange indeed

**Reply by barbudor** [reply]
@dachshund-digital 
Your logs don't show enough information about your setup. We have to trust you that your setup is correct while you don't trust Tasmota for doing the right things.
I imagine that `broker.dd.org` would resolve only from inside your network as it doesn't resolve from here ?
It doesn't resolve neither when using your Internet exposed DNS (through port ***.***.***.***:53)

What is the result for command `IpAddress` ?
Is `IpAddress4` actually pointing to your local DNS cache ?

Some captures with tcpdump on the DNS cache host may also help you to figure out if Tasmota is actually trying to use your local DNS cache.

Works fine for me
```
10:16:05.006 MQT: Attempting connection...
10:16:05.031 MQT: Connected
...
10:16:55.480 CMD: mqtthost
10:16:55.489 MQT: stat/nodemcu/MQTTHOST = {"MqttHost":"mqtt.*****.net"}
10:17:05.397 CMD: ping mqtt.*****.net
10:17:05.407 MQT: stat/nodemcu/PING = {"Ping":"Done"}
10:17:06.525 MQT: tele/nodemcu/PING = {"Ping":{"mqtt.*****.net":{"Reachable":true,"IP":"192.168.***.***","Success":1,"Timeout":0,"MinTime":8,"MaxTime":8,"AvgTime":8}}}
```

**Comment by dachshund-digital** [comment]
I am not sure that my environment is not at fault.  Thinking about it, if tasmota had a hidden issue, it should have shown up well before I came along.  :)  I cannot be the first or only person to use enterprise DNS methods in a home lab, right?

DNS has a way of surprising even those of us that work with it in detail.  I am going to change the configuration of the A, AAA and CNAME records.  And do some more testing.  I have several hosts that are multi-homed, i.e. they have Ethernet and WiFi interfaces... separate IPs since I am not using bonding (or even bridging).  And the broker is one such.  I am going to remove the multi-homed setup and see if that resolves the issue as well. 

I have noticed  that depending on which API calls are used for some OSes, that the order of IP addresses received can be a surprise, some environments get 192.168.1.10 then 192.168.1.20 for a given host, but other environments get 192.168.1.20, then 192.168.1.10 from the same DNS server.  If Tasmota only handles the first IP returned, rather than a IP list?  That would be on example of a surprise... typically you only handle the first resolved result.  Windows 10 versus some Linux distros... this happens for example.

As you noted, the next step would be tcpdumps and maybe flush the DNS cache.  This will take a few days to get through, but I believe you when you say name resolving should be working for 'straight' DNS versus mDNS or even SD-DNS.

**Reply by barbudor** [reply]
Tasmota is using `Wifi.hostbyName()` call from the ES8266 Arduino core which returns a single IP address, not a list.
So Tasmota doesn't get a list to go through to establish a connection. The 1st and only IP address must be the right one.

I always found multi-homing in the same subnet to be source of problems.

**Comment by dachshund-digital** [comment]
Yeah, that is it.  I just happened to look at my original ESP code, and I did not recall that I... wrote logic to handle a list of returned IP addresses.  Nuts.  Moving my ESP devices off my custom doe to Tasmota has been fun, but tripped up.  Yes, multi-homed is tricky.  I am looking at using bonding, so hide the multiple interfaces under a single IP.  This is straight forward with Ethernet NICs, but Ethernet and WiFI has some pitfalls.  Well, now things make sense, so either I use just an IP or limit the broker to a single interface.  Thanks for the assistance.  Still love Tasmota, the more I use it, the better I like it.  A bit more tweaking and I should have Tasmota doing everything my ESP C code was doing!  Consider this issue closed.
